libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+        emit_ev("failure.report_close", {
+            {"failure", "library_error"},
+            {"library_error_context", err},
+        });
+      } else {
+        emit_ev("status.report_close", {{"report_id", ctx.report_id}});
+      }
+    }
+    emit_ev("status.progress", {{"percentage", 1.0},
+                                {"message", "report close"}});
+  } while (0);
+  // TODO(bassosimone): decide whether it makes sense to have an overall
+  // precise error code in this context (it seems not so easy). For now just
+  // always report success, which is what also legacy MK code does.
+  emit_ev("status.end", {{"failure", ""},
+                         {"downloaded_kb", info.bytes_down.load() / 1024.0},
+                         {"uploaded_kb", info.bytes_up.load() / 1024.0}});
+  return true;
+}
+
+void Runner::interrupt() noexcept { interrupted_ = true; }
+
+LogLevel Runner::get_log_level() const noexcept { return settings_.log_level; }
+
+// Methods you typically want to override
+// ``````````````````````````````````````
+
+void Runner::on_event(const nlohmann::json &event) const noexcept {
+  // When running with -fsanitize=thread enable on macOS, a data race in
+  // accessing std::clog is reported. Attempt to avoid that.
+  static std::mutex no_data_race;
+  std::unique_lock<std::mutex> _{no_data_race};
+  std::clog << event.dump() << std::endl;
+}
+
+// Methods you generally DON'T want to override
+// ````````````````````````````````````````````
+
+void Runner::emit_ev(std::string key, nlohmann::json value) const noexcept {
+  assert(value.is_object());
+  on_event({{"key", std::move(key)}, {"value", std::move(value)}});
+}
+
+bool Runner::run_with_index32(
+    const std::chrono::time_point<std::chrono::steady_clock> &begin,
+    const std::string &test_start_time,
+    const std::vector<std::string> &inputs, const NettestContext &ctx,
+    const std::string &collector_base_url, uint32_t i,
+    BytesInfo *info) const noexcept {
+  if (info == nullptr) return false;
+  {
+    auto current_time = std::chrono::steady_clock::now();
+    std::chrono::duration<double> elapsed = current_time - begin;
+    // We call a nettest done when we reach 90% of the expected runtime. This
+    // accounts for possible errors and for the time for closing the report.
+    if (settings_.max_runtime >= 0 &&
+        elapsed.count() >= settings_.max_runtime * 0.9) {
+      LIBNETTEST2_EMIT_INFO("exceeded max runtime");
+      return false;
+    }
+  }
+  emit_ev("status.measurement_start", {{"idx", i}, {"input", inputs[i]}});
+  nlohmann::json measurement;
+  measurement["annotations"] = settings_.annotations;
+  measurement["annotations"]["engine_name"] = settings_.engine_name;
+  measurement["annotations"]["engine_version"] = settings_.engine_version;
+  measurement["annotations"]["engine_version_full"] = settings_.engine_version;
+  measurement["annotations"]["platform"] = LIBNETTEST2_PLATFORM;
```

Comment:
If we also want to support flagging lepidotper (or other custom embedded device platforms) maybe should allow overriding this options. otoh, maybe it's cleaner to do that inside of another annotation key.

Reply:
So, this is probably a good moment to reason about our taxonomy. I understand it like this:

- `engine` is `libmeasurement_kit` (of which this library is just a tiny subset)

- `platform` is the operating system platform

- `application` is something like `OONI` or `measurement_kit`

What is `Lepidopter` in this context? The Linux distro?

Reply:
I guess `lepidopter` would be `distro` in this case. I think what you call above `application` is really `software_name`.

That said the reason to track `lepidopter` in the measurements is that we wanted to monitor the number of deployments and be able to distinguish the measurements from that  type of deployment from other linux flavour.
Since measurements from `lepidopter` have been historically writing the `platform` key, I think that if we do those deployments in the future we should do it inside of the same key for consistency.

Reply:
> I think what you call above application is really software_name.

Correct.

> Since measurements from `lepidopter` [...]

I see. Okay, than basically it would be nice to be able to override the platform.

Reply:
See f1fa00a017e48f61ff03e6900fb7958b1b5c53e3